src/libraries/Common/src/Internal/Cryptography/HashProviderCng.cs

```diff
@@ -105,9 +92,27 @@ public override bool TryFinalizeHashAndReset(Span<byte> destination, out int byt
                 throw Interop.BCrypt.CreateCryptographicException(ntStatus);
             }
 
-            bytesWritten = _hashSize;
             ResetHashObject();
-            return true;
+            return _hashSize;
+        }
+
+        public override int GetCurrentHash(Span<byte> destination)
+        {
+            Debug.Assert(destination.Length >= _hashSize);
+
+            Debug.Assert(_hHash != null);
+
+            using (SafeBCryptHashHandle tmpHash = Interop.BCrypt.BCryptDuplicateHash(_hHash))
```

Comment:
Why is the handle duplication required?

Reply:
It doesn't really duplicate a handle in the same sense that `DuplicateHandle` does where you have a new handle referring to the same object. It duplicates the state of the hash and returns a new handle to it.

CNG gives you an opaque handle to a hash object that carries the state of the hash, and that state reflects the data that has been appended to the hash (`BCryptHashData`). In order to get the final hash, the hash needs to be finalized (`BCryptFinishHash`), but after that you can't append more data.

To support "I want the hash, but I also want to append to more later" the state gets duplicated, one of them is finalized to get the hash, and then you can keep appending to the other handle.

Reply:
Yep, what he said.

More information than you probably want:

Without loss of generality, the finalize step writes the length (as bytes) into the mix so that `{ 0, 0 }` and `{ 0 }` produce different hashes (since the math was on `{ 0, 0, 2, 0, 0, 0 }` and `{ 0, 1, 0, 0, 0 }`), which is destructive to the hash state. 

Since we want to resume, we need to copy the state before the length gets appended, then finish off one of the two.

Reply:
Thanks for the explanations.